src/generic/AhoCorasick.jl

```diff
@@ -0,0 +1,201 @@
+AhoCorasickMatch(6, 1, [1, 2, 3, 4])
+
+
+```
+""" mutable struct AhoCorasickAutomaton
+    goto::Vector{Dict{Int,Int}}
+    fail::Vector{Int}
+    """
+    Output stores for each node a tuple (i, k), where i is the index of the keyword k in
+    the original list of keywords. If several keywords would be the output of the node, only
+    the one with the smallest index is stored
+    """
+    output::Vector{Tuple{Int,Word}}
+end
+
+Markdown.@doc doc"""
+    AhoCorasickMatch(last_position::Int, keyword_index::Int, keyword::Vector{Int})
+
+The return value of searching in a given word with an AhoCorasickAutomaton. Contains the position of the last letter in
+the word that matches a keyword in the automaton, an index of the keyword that was matched and the keyword itself.
+
+# Examples:
+```jldoctest; setup = :(using AbstractAlgebra.Generic)
+julia> keywords = [[1, 2, 3, 4], [1, 5, 4], [4, 1, 2], [1, 2]]
+julia> aut = aho_corasick_automaton(keywords)
+julia> search(aut, [10, 4, 1, 2, 3, 4])
+AhoCorasickMatch(6, 1, [1, 2, 3, 4])
+
+julia> search(aut, [10, 4, 1, 2, 3, 4]).last_position
+6
+
+julia> search(aut, [10, 4, 1, 2, 3, 4]).keyword_index
+1
+
+julia> search(aut, [10, 4, 1, 2, 3, 4]).keyword
+4-element Vector{Int64}:
+ 1
+ 2
+ 3
+ 4
+
+""" struct AhoCorasickMatch
+    last_position::Int
+    keyword_index::Int
+    keyword::Word
+end
+
+function aho_corasick_match(last_position::Int, keyword_index::Int, keyword::Word)
+    return AhoCorasickMatch(last_position, keyword_index, keyword)
+end
+
+Base.hash(m::AhoCorasickMatch) =
+    hash(m.last_position, hash(m.keyword_index, hash(m.keyword)))
+function ==(m1::AhoCorasickMatch, m2::AhoCorasickMatch)
+    return m1.last_position == m2.last_position &&
+           m1.keyword_index == m2.keyword_index &&
+           m1.keyword == m2.keyword
+end
+
+function AhoCorasickAutomaton(keywords::Vector{Word})
+    automaton = AhoCorasickAutomaton([], [], [])
+    construct_goto!(automaton, keywords)
+    construct_fail!(automaton)
+    return automaton
+end
+
+function aho_corasick_automaton(keywords::Vector{Word})
+    return AhoCorasickAutomaton(keywords)
+end
+
+function lookup(automaton::AhoCorasickAutomaton, current_state::Int, next_letter::Int)
+    ret_value = get(automaton.goto[current_state], next_letter, nothing)
+    if current_state == 1 && isnothing(ret_value)
+        return 1
+    end
+    return ret_value
+end
+
+
+function Base.length(automaton::AhoCorasickAutomaton)
+    return length(automaton.goto)
+end
+
+function new_state!(automaton)
+    push!(automaton.goto, Dict{Int,Int}())
+    push!(automaton.output, (typemax(Int), []))
+    push!(automaton.fail, 1)
+    return length(automaton.goto)
+end
+
+function enter!(automaton::AhoCorasickAutomaton, keyword::Word, current_index)
+    current_state = 1
+    for c in keyword
+        new_state = get!(automaton.goto[current_state], c) do
+            new_state!(automaton)
+        end
+        current_state = new_state
```

Comment:
Now that you changed it to use `get!`, you can change it further ;-)
```suggestion
        current_state = get!(automaton.goto[current_state], c) do
            new_state!(automaton)
        end
```